File: portal/backend/app/schemas/designer.py
"""Visual IaC Designer Schemas"""

from pydantic import BaseModel, Field
from typing import List, Dict, Any, Optional
from datetime import datetime
from enum import Enum


class ProjectStatus(str, Enum):
    """Project status enum"""
    DRAFT = "draft"
    GENERATING = "generating"
    GENERATED = "generated"
    DEPLOYING = "deploying"
    DEPLOYED = "deployed"
    FAILED = "failed"


class DiagramNode(BaseModel):
    """Diagram node (component)"""
    id: str
    type: str
    position: Dict[str, float]  # {x: 0, y: 0}
    data: Dict[str, Any]  # Component configuration
    style: Optional[Dict[str, Any]] = None


class DiagramEdge(BaseModel):
    """Diagram edge (connection)"""
    id: str
    source: str  # Source node ID
    target: str  # Target node ID
    type: str = "default"
    data: Optional[Dict[str, Any]] = None
    style: Optional[Dict[str, Any]] = None


class DiagramData(BaseModel):
    """Complete diagram data"""
    nodes: List[DiagramNode] = []
    edges: List[DiagramEdge] = []
    viewport: Optional[Dict[str, Any]] = None


class ProjectBase(BaseModel):
    """Base project schema"""
    name: str = Field(..., min_length=1, max_length=200)
    description: Optional[str] = None
    environment: str = "development"
    deployment_platform: str = "spacelift"


class ProjectCreate(ProjectBase):
    """Create project request"""
    diagram: Optional[DiagramData] = None
    configuration: Optional[Dict[str, Any]] = None


class ProjectUpdate(BaseModel):
    """Update project request"""
    name: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = None
    diagram: Optional[DiagramData] = None
    configuration: Optional[Dict[str, Any]] = None
    environment: Optional[str] = None
    deployment_platform: Optional[str] = None


class ProjectResponse(ProjectBase):
    """Project response"""
    id: int
    created_by_id: int
    diagram: Dict[str, Any]
    configuration: Dict[str, Any]
    generated_terraform: Dict[str, Any]
    generated_pipeline: Dict[str, Any]
    status: str
    import_source: Optional[str]
    import_filename: Optional[str]
    landing_zone_id: Optional[int]
    created_at: datetime
    updated_at: Optional[datetime]
    generated_at: Optional[datetime]
    deployed_at: Optional[datetime]

    class Config:
        from_attributes = True


class ProjectList(BaseModel):
    """Paginated project list"""
    items: List[ProjectResponse]
    total: int
    skip: int
    limit: int


class GenerateTerraformRequest(BaseModel):
    """Generate Terraform request"""
    validate: bool = True
    format: bool = True


class GenerateTerraformResponse(BaseModel):
    """Generated Terraform response"""
    files: Dict[str, str]  # filename: content
    validation_errors: List[str] = []
    warnings: List[str] = []


class GeneratePipelineRequest(BaseModel):
    """Generate pipeline request"""
    platform: str = "spacelift"  # spacelift or github_actions


class GeneratePipelineResponse(BaseModel):
    """Generated pipeline response"""
    configuration: Dict[str, Any]
    files: Dict[str, str]  # filename: content


class ComponentTemplateResponse(BaseModel):
    """Component template response"""
    id: int
    type: str
    name: str
    category: str
    icon: Optional[str]
    terraform_module: Optional[str]
    schema: Dict[str, Any]
    default_config: Dict[str, Any]
    ui_config: Dict[str, Any]

    class Config:
        from_attributes = True


class ImportDiagramRequest(BaseModel):
    """Import diagram request"""
    source: str  # lucidchart, drawio, json
    content: str  # Base64 encoded file content or JSON string
    filename: str


class ImportDiagramResponse(BaseModel):
    """Import diagram response"""
    project_id: int
    diagram: DiagramData
    warnings: List[str] = []


class ValidationError(BaseModel):
    """Validation error"""
    node_id: str
    field: str
    message: str
    severity: str = "error"  # error, warning


class ValidationResponse(BaseModel):
    """Validation response"""
    valid: bool
    errors: List[ValidationError] = []
    warnings: List[ValidationError] = []


class DeployRequest(BaseModel):
    """Deploy project request"""
    auto_approve: bool = False
    environment: str = "development"

File: portal/frontend/src/api/designer.ts
import apiClient from './client'

export interface DiagramNode {
  id: string
  type: string
  position: { x: number; y: number }
  data: Record<string, any>
  style?: Record<string, any>
}

export interface DiagramEdge {
  id: string
  source: string
  target: string
  type?: string
  data?: Record<string, any>
  style?: Record<string, any>
}

export interface DiagramData {
  nodes: DiagramNode[]
  edges: DiagramEdge[]
  viewport?: { x: number; y: number; zoom: number }
}

export interface Project {
  id: number
  name: string
  description: string
  created_by_id: number
  diagram: DiagramData
  configuration: Record<string, any>
  generated_terraform: Record<string, string>
  generated_pipeline: Record<string, any>
  status: string
  environment: string
  deployment_platform: string
  import_source?: string
  import_filename?: string
  landing_zone_id?: number
  created_at: string
  updated_at?: string
  generated_at?: string
  deployed_at?: string
}

export interface ComponentTemplate {
  id: number
  type: string
  name: string
  category: string
  icon?: string
  terraform_module?: string
  schema: {
    inputs: Array<{
      name: string
      type: string
      required: boolean
      description?: string
    }>
    outputs?: Array<{
      name: string
      description?: string
    }>
  }
  default_config: Record<string, any>
  ui_config: {
    width?: number
    height?: number
    color?: string
  }
}

export interface CreateProjectRequest {
  name: string
  description?: string
  environment?: string
  deployment_platform?: string
  diagram?: DiagramData
  configuration?: Record<string, any>
}

export interface UpdateProjectRequest {
  name?: string
  description?: string
  diagram?: DiagramData
  configuration?: Record<string, any>
  environment?: string
  deployment_platform?: string
}

export interface GenerateTerraformResponse {
  files: Record<string, string>
  validation_errors: string[]
  warnings: string[]
}

export interface ValidationResponse {
  valid: boolean
  errors: Array<{
    node_id: string
    field: string
    message: string
    severity: string
  }>
  warnings: Array<{
    node_id: string
    field: string
    message: string
    severity: string
  }>
}

export const designerApi = {
  // Projects
  listProjects: async (params?: { skip?: number; limit?: number; status?: string }) => {
    const response = await apiClient.get('/designer/projects', { params })
    return response.data
  },

  getProject: async (id: number) => {
    const response = await apiClient.get(`/designer/projects/${id}`)
    return response.data
  },

  createProject: async (data: CreateProjectRequest) => {
    const response = await apiClient.post('/designer/projects', data)
    return response.data
  },

  updateProject: async (id: number, data: UpdateProjectRequest) => {
    const response = await apiClient.patch(`/designer/projects/${id}`, data)
    return response.data
  },

  deleteProject: async (id: number) => {
    const response = await apiClient.delete(`/designer/projects/${id}`)
    return response.data
  },

  // Code Generation
  generateTerraform: async (id: number, validate: boolean = true) => {
    const response = await apiClient.post<GenerateTerraformResponse>(
      `/designer/projects/${id}/generate/terraform`,
      { validate, format: true }
    )
    return response.data
  },

  validateProject: async (id: number) => {
    const response = await apiClient.post<ValidationResponse>(
      `/designer/projects/${id}/validate`
    )
    return response.data
  },

  // Components
  listComponents: async (category?: string) => {
    const response = await apiClient.get('/designer/components', {
      params: category ? { category } : undefined,
    })
    return response.data
  },

  getComponent: async (type: string) => {
    const response = await apiClient.get(`/designer/components/${type}`)
    return response.data
  },
}
